### ISSUE TITLE
[CI][Python] Upload Hypothesis example database as artifact

### ISSUE THREAD

**Issue by pitrou** [apache/arrow#41214]
### Describe the enhancement requested

Hypothesis tests use randomly-generated input data. When the data triggers a crash such as [here](https://github.com/ursacomputing/crossbow/actions/runs/8682389434/job/23806853642#step:7:5562), hypothesis to print out the seed for reproduction purposes.

In these cases, it may be useful for the [example database](https://hypothesis.readthedocs.io/en/latest/database.html) to be uploaded as a CI artifact. Presumably, that would allow for local reproduction by a developer.

### Component(s)

Continuous Integration, Python

**Comment by pitrou** [comment]
@jorisvandenbossche @raulcd

**Comment by pitrou** [comment]
In the aforementioned CI run, though, the crash seems to occur when generating the input data itself, so obviously it could not have been written out to the example database.

**Comment by jorisvandenbossche** [comment]
From a previous time that I looked into some hypothesis failures, I seem to remember that even when the crash only happens during the test itself and not before, it does not print information or add the example to the database (in case of a _crash_ and not a normal test failure). 
At least that's what I wrote in https://github.com/HypothesisWorks/hypothesis/issues/3487 a while ago. The suggestion from that issue is to run the tests in a subprocess, which I suppose might help for crashes during a test, but maybe still not for crashes during data generation like the example from this issue.
